src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIAssistantClient.cs

```diff
@@ -0,0 +1,379 @@
+    {
+        // Extract necessary state from chatMessages and options.
+        (string? assistantId, string? threadId) = GetIds(options);
+        (RunCreationOptions runOptions, List<FunctionResultContent>? toolResults) = CreateRunOptions(chatMessages, options);
+
+        // Ensure we have an assistant, finding one if none was specified, and creating one if none was available.
+        if (assistantId is null)
+        {
+            await foreach (var assistant in _assistantClient.GetAssistantsAsync(cancellationToken: cancellationToken).ConfigureAwait(false))
+            {
+                assistantId = assistant.Id;
+                break;
+            }
+
+            if (assistantId is null)
+            {
+                if (options?.ModelId is string model)
+                {
+                    var assistant = await _assistantClient.CreateAssistantAsync(model, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    assistantId = assistant.Value.Id;
+                }
+                else
+                {
+                    Throw.ArgumentException(nameof(options), "No assistant was available, and no model was supplied for which to create an assistant.");
+                }
+            }
+        }
+
+        // Ensure we have a thread.
+        if (threadId is null)
+        {
+            if (toolResults is not null)
+            {
+                Throw.ArgumentException(nameof(chatMessages), "No thread ID was provided, but chat messages includes tool results.");
+            }
+            else if (options is null)
+            {
+                Throw.ArgumentException(nameof(options), "No thread ID was provided, and no options was supplied to store a new thread's ID into.");
+            }
+        }
+
+        // Get the updates to process from the assistant. If we have any tool results, this means submitting those and ignoring
+        // our runOptions. Otherwise, create a run, and a thread if we don't have one.
+        IAsyncEnumerable<StreamingUpdate> updates;
+        if (toolResults?.Select(frc => frc.Context as string).OfType<string>().FirstOrDefault() is string runId)
+        {
+            updates = _assistantClient.SubmitToolOutputsToRunStreamingAsync(
+                threadId, runId,
+                toolResults.Where(frc => frc.Context as string == runId).Select(frc => new ToolOutput(frc.CallId, frc.Result?.ToString() ?? string.Empty)), cancellationToken);
+        }
+        else if (threadId is null)
+        {
+            var creationOptions = new ThreadCreationOptions();
+            foreach (var message in runOptions.AdditionalMessages)
+            {
+                creationOptions.InitialMessages.Add(message);
+            }
+
+            runOptions.AdditionalMessages.Clear();
+
+            updates = _assistantClient.CreateThreadAndRunStreamingAsync(assistantId, creationOptions, runOptions, cancellationToken: cancellationToken);
```

Comment:
Similarly here. It's creating a new thread and storing the ID back into ChatOptions. Not sure what would be better.

Reply:
It is a bit quirky. Maybe I've forgotten but is this the first case where we'd mutate `ChatOptions` automatically? It might not compose well with middleware since middleware may use `options.Clone()`.

The alternative that follows our existing patterns would be to put the ID onto the resulting `ChatCompletion`/`StreamingChatCompletionUpdate`. Developers would then assign this to their `ChatOptions` manually, which is the equivalent to appending the resulting message to a `List<ChatMessage>`.

Reply:
> The alternative that follows our existing patterns would be to put the ID onto the resulting ChatCompletion/StreamingChatCompletionUpdate

That doesn't work well with middleware like function invocation unless we somehow standardize that pattern.

As with assistant ID, I could also revert here back to what I had initially, which was requiring a thread I'd to be specified.